sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/ExecutorResourceMapper.java

```diff
@@ -171,23 +147,40 @@ public ExecutorResourceMapper(
     }
 
     private OfferEvaluationStage newUpdateEvaluationStage(ResourceLabels resourceLabels) {
-        return toEvaluationStage(resourceLabels.resourceSpec, Optional.of(resourceLabels.resourceId),
-                resourceLabels.persistenceId);
+        return toEvaluationStageForExisting(
+                resourceLabels.getResourceSpec(),
+                Optional.of(resourceLabels.getResourceId()),
+                resourceLabels.getPersistenceId(),
+                resourceLabels.getSourceRoot());
     }
 
     private OfferEvaluationStage newCreateEvaluationStage(ResourceSpec resourceSpec) {
-        return toEvaluationStage(resourceSpec, Optional.empty(), Optional.empty());
+        return toEvaluationStageForNew(resourceSpec);
     }
 
-    private OfferEvaluationStage toEvaluationStage(
+    private OfferEvaluationStage toEvaluationStageForExisting(
             ResourceSpec resourceSpec,
             Optional<String> resourceId,
-            Optional<String> persistenceId) {
+            Optional<String> persistenceId,
+            Optional<String> sourceRoot) {
         if (resourceSpec instanceof VolumeSpec) {
-            return new VolumeEvaluationStage(
-                    (VolumeSpec) resourceSpec, null, resourceId, persistenceId, useDefaultExecutor);
+            return VolumeEvaluationStage.getExisting(
+                    (VolumeSpec) resourceSpec,
+                    null,
+                    resourceId,
+                    persistenceId,
+                    sourceRoot,
+                    useDefaultExecutor);
         } else {
             return new ResourceEvaluationStage(resourceSpec, resourceId, null);
         }
     }
+
+    private OfferEvaluationStage toEvaluationStageForNew(ResourceSpec resourceSpec) {
```

Comment:
Given that this is only called by newCreateEvaluationStage(), its content could be folded into there.